### PR TITLE
feat(polkadot): pallet_assets.account balance resolver for asset hub tokens

### PR DESCRIPTION
**Pull request by gomesalexandre**
closes vultisig/vultisig-sdk#562 (depends on it - stacked on top)

## what

PR A (#562) registered USDT (asset_id=1984) and USDC (asset_id=1337) in knownTokens, but left an interim guard in the balance resolver (`if (input.id) return 0n`) to prevent the resolver from returning the DOT relay-chain balance under a USDT token row.

This PR (B) wires the real query: `pallet_assets.Account` storage on Asset Hub (parachain 1000), so USDT/USDC holders see their actual balance.

### storage key encoding (Substrate spec)

```
twox128("Assets")              // 16 bytes - module hash
+ twox128("Account")           // 16 bytes - storage item hash
+ blake2_128_concat(le_u32(asset_id))  // 16 + 4 bytes
+ blake2_128_concat(account_id_32)     // 16 + 32 bytes
```

Total: 100 bytes = 202 hex chars with `0x` prefix.

`twox128` hashes are hardcoded constants (same pattern as the existing `systemAccountPrefix`) to avoid pulling `@polkadot/util-crypto` — that package's BN.js dep double-bundles in browser/extension builds and crashes at module init (see comment in polkadot.ts for the full story).

`blake2_128` uses `@noble/hashes/blake2b` with `dkLen: 16` — already in the tree.

### response decoding

The response is a SCALE-encoded `AssetAccount`. Only the first 16 bytes matter: `balance: u128` LE. Null response = account has no entry for that asset = 0n.

### Asset Hub RPC

Added `assetHubRpcUrl = ${rootApiUrl}/dot-ah/` to `polkadot/client.ts`. Token queries go there; native DOT relay-chain queries still use the existing `polkadotRpcUrl = ${rootApiUrl}/dot/`.

## storage key verification

USDT (asset_id=1984) + Alice pubkey (`d43593c...`):

```
twox128("Assets")  = 682a59d51ab9e48a8c8cc418ff9708d2
twox128("Account") = b99d880ec681799c0cf30e8886371da9
le_u32(1984)       = c0070000
blake2_128(le_u32(1984))  = a319d0e87221ca1ee751c1529f201522
blake2_128(alice_pubkey)  = de1e86a9a8c739864cf3cc5ec2bea59f

full key = 0x682a59d51ab9e48a8c8cc418ff9708d2
             b99d880ec681799c0cf30e8886371da9
             a319d0e87221ca1ee751c1529f201522c0070000
             de1e86a9a8c739864cf3cc5ec2bea59f
             d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d
           (202 chars - matches Substrate spec)
```

Verified against `@polkadot/util-crypto` (`xxhashAsHex` + `blake2AsHex`) offline, and cross-checked that `@noble/hashes/blake2b` with `dkLen: 16` produces identical output.

USDC (asset_id=1337) similarly verified (`le_u32(1337)=39050000`).

## tests (8 total, replaces 5 interim-guard tests)

| test | what it covers |
|---|---|
| USDT balance fetch | correct balance decode + storage key shape + `dot-ah` URL |
| USDC balance fetch | correct balance decode + `le_u32(1337)` in key |
| missing account (null) | → 0n |
| malformed SCALE response | → throws `unexpected storage response` |
| non-numeric asset_id | → throws `Invalid Asset Hub asset_id` (no RPC call) |
| Asset Hub RPC error | → propagates error |
| native DOT: relay chain URL | confirms `dot-ah` NOT used, `System.Account` prefix used |
| native DOT: null account | → 0n |

## stacking note

This PR is based on `feat_register_asset_hub_usdt_usdc` (#562). Review/merge order: #562 first, then this.

## risk

Low - isolated to the Polkadot balance resolver. Native DOT path unchanged (confirmed by tests checking URL + storage key prefix). No other chain touched.

no-receipts-required (pure resolver + no user-facing surface change, SDK-level only)

🤖 Agent-generated

### CODE REVIEW

**Review by gomesalexandre — COMMENTED** [review]
ser, this one got the full treatment — multi-lane Claude correctness + live mainnet RPC cross-check. here's the synthesis.

---

## verdict: APPROVE (with `should-fix` + suggestions)

zero blockers. no fund-safety issues. storage key encoding is correct, live cross-check passed, native DOT path is untouched.

---

## lane 1 — storage key correctness (PASSED, live-verified)

ran `state_getKeysPaged` against `polkadot-asset-hub-rpc.polkadot.io` to pull real USDT (asset_id=1984) account entries, then queried one of them directly:

```
response: 0x0a3a00000000000000000000000000000001
decoded:  14858 USDT micro-units (= 0.014858 USDT @ 6 dec)
```

decoder produced the correct value. the key structure was verified byte-by-byte:

```
0x682a59d51ab9e48a8c8cc418ff9708d2  <- twox128("Assets")      ✓
  b99d880ec681799c0cf30e8886371da9  <- twox128("Account")     ✓ (same as system prefix)
  a319d0e87221ca1ee751c1529f201522  <- blake2_128(le_u32(1984)) ✓
  c0070000                          <- le_u32(1984)            ✓
  00064a5b7db2b0c8271fe85bce78da1e  <- blake2_128(account)    ✓
  08622ff9ded...                    <- raw account pubkey      ✓
total: 202 chars                                               ✓
```

`le_u32(1984) = c0070000` and `le_u32(1337) = 39050000` both confirmed correct. the `blake2_128_concat` spec (hash + raw) is implemented correctly.

---

## lane 2 — SCALE decode

- reads `hex.slice(0, 32)` (16 bytes) as LE u128 - correct
- reverses byte-pairs for BE bigint construction - correct
- null → 0n - correct
- `hex.length < 32` guard - correct
- live response has 36 chars (18 bytes: balance + status + reason enum) — parser correctly ignores trailing bytes ✓

---

## lane 3 — RPC URL routing

`assetHubRpcUrl = rootApiUrl + '/dot-ah/'` — confirmed this is a constant string (not env-dependent). native DOT still routes to `polkadotRpcUrl = rootApiUrl + '/dot/'`. tests verify the URL routing explicitly. ✓

---

## lane 4 — test suite (8 tests, all green)

ran `yarn vitest run --config .config/vitest.core.config.ts packages/core/chain/coin/balance/resolvers/polkadot.test.ts` — **8/8 pass**.

note: running the file directly without the config alias (`yarn vitest run <file>`) produces a false failure because `@vultisig/core-chain/...` isn't resolved, making `assetHubRpcUrl` undefined and the URL assertion fail. CI uses `yarn test:core` which hits the correct config, so this is not a CI issue — just a footgun for local ad-hoc runs.

---

## findings

### should-fix: test fixture `freeLE` is 1 byte too long

`polkadot.test.ts:153`:

```ts
const freeLE = '0010A5D4E8000000000000000000000000' // 34 chars = 17 bytes
```

`u128` = 16 bytes = 32 hex chars. this fixture has 17 bytes. the test **still passes** because `hex.slice(32, 64)` in the decoder reads exactly the first 16 bytes of `freeLE` — the extra trailing `00` is never consumed. but the comment says `// free (16 bytes LE)` and the constant is wrong. it'll confuse the next person who reads it, and a non-zero 17th byte would silently corrupt the `reserved` field offset, potentially masking bugs in future AccountInfo layout changes.

fix: trim to 32 chars: `'0010A5D4E800000000000000000000000000'` → `'0010A5D4E8000000000000000000000000'`

wait, let me count again for you: `1_000_000_000_000 = 0xE8D4A51000`, LE 16 bytes = `[00, 10, A5, D4, E8, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00]` = `'0010A5D4E80000000000000000000000'` (32 chars). drop the trailing `00`.

---

### suggestion: asset_id validation misses truncating-parse cases

`polkadot.ts:68`:

```ts
const assetId = parseInt(assetIdStr, 10)
if (!Number.isInteger(assetId) || assetId <= 0) {
```

three edge cases pass the guard silently:

- `'1984.5'` → `parseInt` returns `1984` (truncates) — still queries the right asset, but input was malformed
- `'1984abc'` → `parseInt` returns `1984` — same story
- `'4294967296'` → `parseInt` returns `4294967296`, `Number.isInteger` returns `true`, but `DataView.setUint32` wraps to `0` silently — this would query asset_id=0

in practice `coin.id` only ever comes from `knownTokens` (exact strings `'1984'`, `'1337'`) so none of these bite today. but if `coin.id` ever accepts user input, `'4294967296'` silently querying asset_id=0 would be confusing. suggestion: replace `parseInt` with a strict numeric check:

```ts
const assetId = Number(assetIdStr)
if (!Number.isInteger(assetId) || assetId <= 0 || assetId > 0xFFFFFFFF) {
  throw new Error(`Invalid Asset Hub asset_id: ${assetIdStr}`)
}
```

`Number('1984abc')` = `NaN`, `Number('1984.5')` = `1984.5` (not integer), `Number('4294967296')` = 4294967296 (caught by > 0xFFFFFFFF). also add a test case for asset_id=0 and the overflow case.

---

### suggestion: `assetHubRpcUrl` not mocked in the existing `getPolkadotClient` - minor isolation note

`client.ts` now exports both `polkadotRpcUrl` and `assetHubRpcUrl`, but `getPolkadotClient` still constructs its `ApiPromise` from `polkadotRpcUrl` (relay chain). if anyone ever wires `getPolkadotClient` for Asset Hub queries instead of the raw `queryUrl` approach, they'll get the wrong endpoint. a comment on `getPolkadotClient` noting it's relay-chain only would preempt that footgun. non-blocking.

---

### q: CI still `pending` at review time

`gh pr checks 563` showed `pending` for all three CI jobs. this PR is a draft so that's probably fine, but worth noting that the 8-test green is confirmed locally via `yarn test:core`, not from CI yet.

---

### praise: live-chain verification in PR body

the storage key breakdown in the PR description with `blake2_128(le_u32(1984))` spelled out hex-by-hex is exactly what a future auditor needs. also the comment explaining WHY twox128 hashes are hardcoded (the BN.js double-bundle crash) is the right level of institutional memory to preserve.

---

@NeOMakinG this is stacked on #562 — review order: #562 first, then #563 rebases onto main.

File: packages/core/chain/coin/balance/resolvers/polkadot.test.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    queryUrlMock.mockReset()
+  })
+
+  it('queries relay chain RPC for native DOT (no id present)', async () => {
+    // SCALE-encoded AccountInfo with free balance = 1_000_000_000_000n (1 DOT)
+    // Layout: nonce(u32=0) + consumers(u32=0) + providers(u32=1) + sufficients(u32=0)
+    //         + free(u128) + reserved(u128) + frozen(u128) + flags(u128)
+    // free = 1_000_000_000_000 = 0x000000E8D4A51000 LE-encoded in 16 bytes:
+    //   LE bytes: 00 10 A5 D4 E8 00 00 00 00 00 00 00 00 00 00 00
+    const freeLE = '0010A5D4E8000000000000000000000000'
```
Comment (gomesalexandre) [inline packages/core/chain/coin/balance/resolvers/polkadot.test.ts]: should-fix: `freeLE` fixture is 34 hex chars (17 bytes) but u128 is 16 bytes = 32 chars. the comment says `// free (16 bytes LE)` but the constant has a trailing `00` — test passes today because the decoder reads `slice(32, 64)` so the 17th byte is never consumed. but it'll confuse the next reader and could mask bugs if the AccountInfo layout test is ever extended (non-zero 17th byte silently corrupts the `reserved` field offset in the fixture).

fix: `'0010A5D4E80000000000000000000000'` (32 chars, 16 bytes).

Comment (gomesalexandre) [reply packages/core/chain/coin/balance/resolvers/polkadot.test.ts]: tackled in 3ab72030: dropped trailing `00` from freeLE (34→32 hex chars, 17→16 bytes). also tackled the parseInt→Number() u32 guard + overflow/fractional/negative tests, and added JSDoc on getPolkadotClient in the same commit.

**Review by NeOMakinG — COMMENTED** [review]
gm — multi-lane pass: . conceptual-only (no runtime this pass). comment-only.

## verdict
COMMENTED

🤖 reviewed via [Claude Code](https://claude.com/claude-code) with codex `gpt-5.3-codex` conceptual pass.

**Review by NeOMakinG — COMMENTED** [review]
## r1 — two should-fix items, dirty state is mechanical

The balance resolver implementation is correct. SCALE key construction, LE decode, Asset Hub vs relay-chain routing, and all test golden vectors verified. The `CONFLICTING` mergeable state is a base-branch artifact (see below) with no real git conflicts.

## testing approach

Static triple-lane review: Claude correctness lane (SCALE spec + storage key verification + LE decode trace), Codex adversarial lane (null/frozen/ED edge cases), GLM chain-edge lane (byte-order, encoding, boundary assumptions). No runtime exercise required — pure resolver with no user-facing surface, `no-receipts-required` declared.

## should-fix

1. **Scientific-notation asset_id slips past the integer guard.**
`Number('1.999e3')` evaluates to `1999`, and `Number.isInteger(1999)` is `true`, so the guard accepts it even though the input string is not a pure integer. This is inconsistent with the explicit `'1984.5'` rejection: a caller passing `'1.9840e3'` gets different behaviour than `'1984.5'` despite both being non-canonical representations of integers. All real `knownTokens` asset IDs are plain decimal strings so practical impact is nil, but the validation contract is misleading.
`action:` add a `/^\d+$/` pre-check before `Number()` — or document in the guard comment that scientific notation is accepted.

2. **No test for `asset_id='0'` (lower u32 boundary).**
The guard allows `0` (`assetId < 0` excludes negatives, not zero). `le_u32(0)` = `00000000`. No test verifies the boundary resolves correctly and produces the expected storage key tail. Substrate pallet_assets does use asset ID 0 on some system chains.
`action:` add a test: `id: '0'`, `mockResolvedValue({ result: null })`, assert `0n` returned and `storageKey.includes('00000000')` at the asset segment.

## blocking

**Dirty base — change PR base to `main` before merging.**

The PR targets `feat_register_asset_hub_usdt_usdc`, which has a post-squash commit (`30e3d8f5`) not present in `main`. GitHub flags this as `CONFLICTING`. A `git merge-tree` of the three PR commits (`cf030a67`, `5d57293d`, `77bb9c76`) against `main` is clean — all three apply on top of `921fa96d` (sdk#562's merge commit, already in `main`) with zero conflicts.

`action:` `gh pr edit 563 --base main` — no rebase needed.

## risk

Low. Resolver is isolated to the Polkadot balance path. Native DOT path unchanged (confirmed by tests asserting relay-chain URL and `System.Account` prefix). No other chain touched.

## verdict

COMMENTED — approach and implementation correct, two should-fix items, merge blocked on base-branch change.

---
*reviewed via [Claude Code](https://claude.com/claude-code) triple-lane (Claude + Codex adversarial + GLM chain-edge). Static analysis only — SDK resolver, no runtime surface.*

File: packages/core/chain/chains/polkadot/client.ts [hunk 2]
```diff
@@ -4,6 +4,14 @@ import { memoizeAsync } from '@vultisig/lib-utils/memoizeAsync'
 
 export const polkadotRpcUrl = `${rootApiUrl}/dot/`
 
+// Asset Hub (parachain 1000) — home of pallet_assets USDT/USDC
+export const assetHubRpcUrl = `${rootApiUrl}/dot-ah/`
```
Comment (NeOMakinG) [inline packages/core/chain/chains/polkadot/client.ts:8]: `should-fix:` Missing Runtime URL Configuration Guard — The `assetHubRpcUrl` is constructed as `${rootApiUrl}/dot-ah/` without validation that the upstream RPC server supports this endpoint. If infrastructure hasn't provisioned the `/dot-ah/` route, all Asset Hub balance queries will fail with 404.

fix: Add environment variable validation for Asset Hub RPC availability before exporting `assetHubRpcUrl`, or document the infrastructure requirement explicitly.

File: packages/core/chain/coin/balance/resolvers/polkadot.ts [hunk 3]
```diff
@@ -51,8 +56,71 @@ const decodePolkadotPublicKey = (address: string): Uint8Array => {
   return payload.subarray(1)
 }
 
+// Storage key: assetsAccountPrefix
```
Comment (NeOMakinG) [inline packages/core/chain/coin/balance/resolvers/polkadot.ts:59]: `should-fix:` Insufficient Asset ID Validation for Future Expansion — The validation `if (!Number.isInteger(assetId) || assetId < 0 || assetId > 0xffffffff)` only checks basic type constraints. Without allowlist validation, any valid u32 asset_id triggers an RPC query, potentially exposing users to malformed balance queries.

fix: Either document that only USDT/USDC are currently supported with an allowlist check, or add metrics/logging for unsupported asset_id queries to monitor usage patterns.

File: packages/core/chain/coin/balance/resolvers/polkadot.ts [hunk 4]
```diff
@@ -51,8 +56,71 @@ const decodePolkadotPublicKey = (address: string): Uint8Array => {
...
+  new DataView(assetIdLE.buffer).setUint32(0, assetId, true)
+
+  const assetIdHashHex = bytesToHex(blake2b(assetIdLE, { dkLen: 16 }))
+  const assetIdHex = bytesToHex(assetIdLE)
+  const accountHashHex = bytesToHex(blake2b(pubkey, { dkLen: 16 }))
+  const accountHex = bytesToHex(pubkey)
+
+  const storageKey = assetsAccountPrefix + assetIdHashHex + assetIdHex + accountHashHex + accountHex
+
+  const response = await queryUrl<RpcResponse<string | null>>(assetHubRpcUrl, {
```
Comment (NeOMakinG) [inline packages/core/chain/coin/balance/resolvers/polkadot.ts:82]: `should-fix:` Missing Error Context for Troubleshooting — When RPC errors occur, the error message lacks critical debugging context: which asset_id was being queried, which address, and the full RPC request payload.

fix: Enhance error messages to include asset_id, address (truncated for privacy), and request details.

File: packages/core/chain/coin/balance/resolvers/polkadot.ts [hunk 5]
```diff
@@ -51,8 +56,71 @@ const decodePolkadotPublicKey = (address: string): Uint8Array => {
...
+      params: [storageKey],
+      id: 1,
+    },
+  })
+
+  if (response.error) {
+    throw new Error(`Asset Hub pallet_assets RPC error: ${response.error.message ?? `code ${response.error.code}`}`)
+  }
+
+  // Null result means the account has no entry for this asset — balance is 0.
```
Comment (NeOMakinG) [inline packages/core/chain/coin/balance/resolvers/polkadot.ts:95]: `should-fix:` Frozen Balance Display Policy Not Enforced — The comment states 'We return the raw u128 balance regardless of account status (Liquid/Frozen/Blocked)' but displaying frozen funds as liquid balance in the UI could mislead users into thinking they can spend locked funds.

fix: Either parse the status byte and return separate liquid/frozen amounts, or add clear JSDoc documentation that frozen balances are intentionally included and consuming UI must handle spend-path validation.

File: packages/core/chain/coin/knownTokens/index.ts [hunk 6]
```diff
@@ -761,6 +761,29 @@ const leanTokens: Partial<LeanChainTokensRecord> = {
...
+    // On-chain symbol for 1984 is "USDt" but we normalise to "USDT" for consistency.
+    '1984': {
+      ticker: 'USDT',
+      logo: 'usdt',
+      decimals: 6,
+      // NOTE: CoinGecko's 'tether' coin doesn't list polkadot-asset-hub in its
+      // platforms map (only Ethereum/Solana/Tron/TON/Near/etc). Price still resolves
+      // correctly via /simple/price?ids=tether but any future platform-verification
+      // logic (e.g. checking coin.platforms['polkadot-asset-hub']) will find nothing
+      // and should special-case this entry.
```
Comment (NeOMakinG) [inline packages/core/chain/coin/knownTokens/index.ts:777]: `suggestion:` CoinGecko Platform Gap Documentation — The NOTE about CoinGecko's 'tether' coin lacking 'polkadot-asset-hub' in its platforms map creates a hidden dependency on the /simple/price endpoint working without platform verification.

fix: Consider adding a TODO comment or tracking issue to monitor CoinGecko platform coverage. If CoinGecko adds platform-specific price endpoints, update this entry.

File: packages/core/chain/coin/balance/resolvers/polkadot.test.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+//
+//   USDT key = 0x682a59d51ab9e48a8c8cc418ff9708d2
+//              b99d880ec681799c0cf30e8886371da9
+//              a319d0e87221ca1ee751c1529f201522c0070000
+//              de1e86a9a8c739864cf3cc5ec2bea59f
+//              d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d
+//   (202 chars total including "0x" prefix)
+
+describe('getPolkadotCoinBalance — pallet_assets.Account (Asset Hub tokens)', () => {
+  beforeEach(() => {
```
Comment (NeOMakinG) [inline packages/core/chain/coin/balance/resolvers/polkadot.test.ts:32]: `suggestion:` Test Coverage Gap: Asset Status Variations — Tests only cover Liquid status (0x00) with Consumer reason (0x01). No test cases for Frozen or Blocked account states, which could hide parsing bugs if SCALE layout varies.

fix: Add test cases for Frozen (status=0x01) and Blocked (status=0x02) account states to ensure balance parsing works across all AssetAccount variants.

File: packages/core/chain/chains/polkadot/client.ts (hunk 2 above)
Comment (NeOMakinG) [inline packages/core/chain/chains/polkadot/client.ts:8]: `q:` Infrastructure Readiness — Has the infrastructure team confirmed that the `/dot-ah/` endpoint is provisioned and production-ready? This PR introduces a hard dependency on this route.

fix: Please confirm infrastructure readiness in the PR description or add a link to the infrastructure provisioning task.

File: packages/core/chain/coin/balance/resolvers/polkadot.ts (hunk 3 above)
Comment (NeOMakinG) [inline packages/core/chain/coin/balance/resolvers/polkadot.ts:59]: `q:` Future Asset Hub Token Support Strategy — The comment mentions 'USDT id=1984, USDC id=1337, etc.' implying additional tokens may be added. Is there a process for discovering and registering new Asset Hub tokens?

fix: Clarify the expansion strategy in the resolver documentation or add a TODO comment tracking the future work item.

**Review by NeOMakinG — APPROVED** [review]
gm — multi-lane pass: . runtime-tested on native-dot-balance-resolver. lgtm.

## testing approach
- native-dot-balance-resolver

## should-fix
### 1. Missing Runtime URL Configuration Guard — `packages/core/chain/chains/polkadot/client.ts:8`
The `assetHubRpcUrl` is constructed as `${rootApiUrl}/dot-ah/` without validation that the upstream RPC server supports this endpoint. If infrastructure hasn't provisioned the `/dot-ah/` route, all Asset Hub balance queries will fail with 404.
action: Add environment variable validation for Asset Hub RPC availability before exporting `assetHubRpcUrl`, or document the infrastructure requirement explicitly.

### 2. Insufficient Asset ID Validation for Future Expansion — `packages/core/chain/coin/balance/resolvers/polkadot.ts:59`
The validation `if (!Number.isInteger(assetId) || assetId < 0 || assetId > 0xffffffff)` only checks basic type constraints. Without allowlist validation, any valid u32 asset_id triggers an RPC query, potentially exposing users to malformed balance queries.
action: Either document that only USDT/USDC are currently supported with an allowlist check, or add metrics/logging for unsupported asset_id queries to monitor usage patterns.

### 3. Missing Error Context for Troubleshooting — `packages/core/chain/coin/balance/resolvers/polkadot.ts:82`
When RPC errors occur, the error message lacks critical debugging context: which asset_id was being queried, which address, and the full RPC request payload.
action: Enhance error messages to include asset_id, address (truncated for privacy), and request details.

### 4. Frozen Balance Display Policy Not Enforced — `packages/core/chain/coin/balance/resolvers/polkadot.ts:95`
The comment states 'We return the raw u128 balance regardless of account status (Liquid/Frozen/Blocked)' but displaying frozen funds as liquid balance in the UI could mislead users into thinking they can spend locked funds.
action: Either parse the status byte and return separate liquid/frozen amounts, or add clear JSDoc documentation that frozen balances are intentionally included and consuming UI must handle spend-path validation.

## suggestion
### 1. CoinGecko Platform Gap Documentation — `packages/core/chain/coin/knownTokens/index.ts:777`
The NOTE about CoinGecko's 'tether' coin lacking 'polkadot-asset-hub' in its platforms map creates a hidden dependency on the /simple/price endpoint working without platform verification.
action: Consider adding a TODO comment or tracking issue to monitor CoinGecko platform coverage. If CoinGecko adds platform-specific price endpoints, update this entry.

### 2. Test Coverage Gap: Asset Status Variations — `packages/core/chain/coin/balance/resolvers/polkadot.test.ts:32`
Tests only cover Liquid status (0x00) with Consumer reason (0x01). No test cases for Frozen or Blocked account states, which could hide parsing bugs if SCALE layout varies.
action: Add test cases for Frozen (status=0x01) and Blocked (status=0x02) account states to ensure balance parsing works across all AssetAccount variants.

### 3. Add integration test once Asset Hub RPC is provisioned — `packages/core/chain/coin/balance/resolvers/polkadot.ts:1`
Unit tests verify storage key encoding, but integration test with live Asset Hub RPC would catch endpoint configuration issues earlier.
action: Add an integration test to the QA suite that queries real USDT/USDC balances once infrastructure is ready.

## q
- Infrastructure Readiness
- Future Asset Hub Token Support Strategy

## risk
Low — isolated to Polkadot balance resolver; native DOT path unchanged; comprehensive test coverage; no breaking API changes. Principal concern is infrastructure readiness of /dot-ah/ endpoint (404 is expected until provisioned).

## verdict
APPROVE — ship it 🫡

🤖 reviewed via [Claude Code](https://claude.com/claude-code) with codex `gpt-5.3-codex` adversarial pass. native-dot-balance-resolver.

**Review by NeOMakinG — APPROVED** [review]
Unit Tests pass (12m52s). Scientific-notation guard documented in test. asset_id='0' boundary test added. All 3 should-fixes addressed. LGTM.
